crates/red_knot_python_semantic/src/types.rs

```diff
@@ -471,6 +471,154 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if this type and `other` have no common elements.
+    ///
+    /// Note: This function aims to have no false positives, but might return
+    /// wrong `false` answers in some cases.
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        match (self, other) {
```

Comment:
It might make some cases unnecessary, but then we would have to add `unreachable!(…)`, right?

Note: the shortcut leads to the wrong result if `self = other = Type::Never`, so it would need an additional condition to handle `Never`.